fastn-core/src/auth/emailpassword.rs

```diff
@@ -0,0 +1,371 @@
+        name: String,
+        password: String,
+    }
+
+    let user: UserPayload = serde_json::from_slice(req.body())?;
+
+    tracing::info!("user payload: {:?}", &user);
+
+    let client = db::get_client().await?;
+
+    let res = client
+        .query(
+            "select count(id) from fastn_user where username = $1 or email = $2",
+            &[&user.username, &user.email],
+        )
+        .await
+        .unwrap();
+
+    let user_count: i64 = res[0].get(0);
+
+    if user_count > 0 {
+        // TODO: use correct HTTP status code
+        return fastn_core::http::api_error("user already exists");
+    }
+
+    // encrypt password
+    let salt =
+        argon2::password_hash::SaltString::generate(&mut argon2::password_hash::rand_core::OsRng);
+
+    let argon2 = argon2::Argon2::default();
+
+    let hashed_password = argon2
+        .hash_password(user.password.as_bytes(), &salt)
+        .unwrap()
+        .to_string();
+
+    let user_id = uuid::Uuid::new_v4();
+
+    // TODO: email verification
+    // TODO: password can be null for oauth users. if it's null they should not login using
+    // emailpassword flow
+    let user = upsert_user(
+        &user_id,
+        user.email.as_str(),
+        user.username.as_str(),
+        user.name.as_str(),
+        hashed_password.as_str(),
+    )
+    .await?;
+
+    tracing::info!("fastn_user created. user: {:?}", &user);
+
+    fastn_core::http::api_ok(user)
+}
+
+pub(crate) async fn login(
+    req: &crate::http::Request,
+    next: String,
+) -> fastn_core::Result<fastn_core::http::Response> {
+    if req.method() != "POST" {
+        return Ok(fastn_core::not_found!("invalid route"));
+    }
+
+    #[derive(serde::Deserialize, Debug)]
+    struct Payload {
+        // TODO: add email/username
+        email: String,
+        password: String,
+    }
+
+    let body: Payload = serde_json::from_slice(req.body())?;
+
+    tracing::info!("request payload: {:?}", body);
+
+    let client = db::get_client().await?;
+
+    let user = client
+        .query(
+            "select id, username, email, name, password from fastn_user where email = $1 limit 1",
+            &[&body.email],
+        )
+        .await
+        .unwrap_or(vec![]);
+
+    if user.is_empty() {
+        return fastn_core::http::api_error("invalid payload");
+    }
+
+    let user = user.first().expect("vec of length must have first");
+
+    let password: String = user.get("password");
+    let user_id: uuid::Uuid = user.get("id");
+
+    let parsed_hash = argon2::PasswordHash::new(&password).unwrap();
+
+    let password_match = Argon2::default().verify_password(body.password.as_bytes(), &parsed_hash);
+
+    if password_match.is_err() {
+        return fastn_core::http::api_error("invalid payload");
+    }
+
+    let session_id = uuid::Uuid::new_v4();
+
+    let affected = create_session(&session_id, &user_id).await.unwrap_or(0);
+
+    tracing::info!("session created. Rows affected: {}", &affected);
```

Comment:
Let's do more informative login, for example log the email address, actually user id as email is PII and we do not want to log PIIs unless necessary.